File: AssistantPlus/Application/View/CoreCommandSender/ValueExpression.cs
#pragma warning disable 0,
// ReSharper disable

using Twinning.AssistantPlus;
using Twinning.AssistantPlus.Utility;

namespace Twinning.AssistantPlus.View.CoreCommandSender {

	public record ValueExpression {
	}

	// ----------------

	public record BooleanExpression : ValueExpression {
		public required Boolean Value { get; set; }
	}

	public record IntegerExpression : ValueExpression {
		public required Integer Value { get; set; }
	}

	public record FloaterExpression : ValueExpression {
		public required Floater Value { get; set; }
	}

	public record StringExpression : ValueExpression {
		public required String Value { get; set; }
	}

	public record SizeExpression : ValueExpression {
		public required Floater Count    { get; set; }
		public required Integer Exponent { get; set; }
	}

	public record PathExpression : ValueExpression {
		public required StoragePath Content { get; set; }
	}

	// ----------------

	public static class ValueExpressionHelper {

		#region utility

		public static String MakeString(
			ValueExpression value
		) {
			return value switch {
				BooleanExpression values => $"{ConvertHelper.MakeBooleanToStringOfConfirmationCharacter(values.Value)}",
				IntegerExpression values => $"{ConvertHelper.MakeIntegerToString(values.Value)}",
				FloaterExpression values => $"{ConvertHelper.MakeFloaterToString(values.Value)}",
				StringExpression values  => $"{values.Value}",
				SizeExpression values    => $"{ConvertHelper.MakeFloaterToString(values.Count)}{new[] { "b", "k", "m", "g" }[values.Exponent]}",
				PathExpression values    => $"{values.Content.Emit()}",
				_                        => throw new UnreachableException(),
			};
		}

		#endregion

	}

}

File: Assistant/lib/view/core_command_sender/value_expression.dart
import '/common.dart';
import '/utility/convert_helper.dart';
import '/utility/storage_path.dart';

// ----------------

class ValueExpression {
  ValueExpression(
  );
  @override
  operator ==(other) => other is ValueExpression;
  @override
  get hashCode => Object.hashAll([]);
}

// ----------------

class BooleanExpression extends ValueExpression {
  Boolean value;
  BooleanExpression(
    this.value,
  );
  @override
  operator ==(other) => other is BooleanExpression && this.value == other.value;
  @override
  get hashCode => Object.hashAll([this.value]);
}

class IntegerExpression extends ValueExpression {
  Integer value;
  IntegerExpression(
    this.value,
  );
  @override
  operator ==(other) => other is IntegerExpression && this.value == other.value;
  @override
  get hashCode => Object.hashAll([this.value]);
}

class FloaterExpression extends ValueExpression {
  Floater value;
  FloaterExpression(
    this.value,
  );
  @override
  operator ==(other) => other is FloaterExpression && this.value == other.value;
  @override
  get hashCode => Object.hashAll([this.value]);
}

class StringExpression extends ValueExpression {
  String value;
  StringExpression(
    this.value,
  );
  @override
  operator ==(other) => other is StringExpression && this.value == other.value;
  @override
  get hashCode => Object.hashAll([this.value]);
}

class SizeExpression extends ValueExpression {
  Floater count;
  Integer exponent;
  SizeExpression(
    this.count,
    this.exponent,
  );
  @override
  operator ==(other) => other is SizeExpression && this.count == other.count && this.exponent == other.exponent;
  @override
  get hashCode => Object.hashAll([this.count, this.exponent]);
}

class PathExpression extends ValueExpression {
  StoragePath content;
  PathExpression(
    this.content,
  );
  @override
  operator ==(other) => other is PathExpression && this.content == other.content;
  @override
  get hashCode => Object.hashAll([this.content]);
}

// ----------------

class ValueExpressionHelper {

  // #region utility

  static String makeString(
    ValueExpression value,
  ) {
    return switch (value) {
      BooleanExpression _ => '${ConvertHelper.makeBooleanToStringOfConfirmationCharacter(value.value)}',
      IntegerExpression _ => '${ConvertHelper.makeIntegerToString(value.value)}',
      FloaterExpression _ => '${ConvertHelper.makeFloaterToString(value.value)}',
      StringExpression  _ => '${value.value}',
      SizeExpression    _ => '${ConvertHelper.makeFloaterToString(value.count)}${['b', 'k', 'm', 'g'][value.exponent]}',
      PathExpression    _ => '${value.content.emitGeneric()}',
      _                   => throw UnreachableException(),
    };
  }

  // #endregion

}
